sidecar/internal/server/server.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2021 The Kubernetes-CSI-Addons Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"errors"
+	"net"
+
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+)
+
+// SidecarService is the interface that is required to be implemented so that
+// the SidecarServer can register the service by calling RegisterService().
+type SidecarService interface {
+	// RegisterService is called by the SidecarServer to add a CSI-Addons
+	// service that can handle requests.
+	RegisterService(server grpc.ServiceRegistrar)
+}
+
+// SidecarServer is the gRPC server that listens on an endpoint
+// where the CSI-Addons requests come in.
+type SidecarServer struct {
+	// URL components to listen on the tcp port
+	scheme   string
+	endpoint string
+
+	server   *grpc.Server
+	services []SidecarService
+}
+
+// NewSidecarServer create a new SidecarServer on the given endpoint. The
+// endpoint should be an ip address. Only tcp ports are supported.
+func NewSidecarServer(endpoint string) *SidecarServer {
+	ss := &SidecarServer{}
+
+	if ss.services == nil {
+		ss.services = make([]SidecarService, 0)
+	}
+
+	ss.scheme = "tcp"
+	ss.endpoint = ":" + endpoint
```

Comment:
Removed localhost from here, since controller could not connect. 
Now controller can connect and successfully communicate with sidecar.